docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,546 @@
+
+AWS JSON 1.0 smithy protocol test for `requestCompression` trait:
+```json
+$version: "2.0"
+
+namespace aws.protocoltests.json10
+
+        use aws.protocols#awsJson1_0
+        use smithy.test#httpRequestTests
+        use smithy.test#httpResponseTests
+
+        apply PutWithContentEncoding @httpRequestTests([
+    {
+        id: "SDKAppliedContentEncoding_awsJson1_0"
+        documentation: "Compression algorithm encoding is appended to the Content-Encoding header."
+        protocol: awsJson1_0
+        params: {
+        "data": """
+                NOTE: REMOVED CONTENTS FOR SAKE OF SPACE. TEST CAN BE FOUND IN REFERENCES SECTION 
+                """
+        }
+        method: "POST"
+        uri: "/"
+        headers: {
+        "Content-Encoding": "gzip"
+        }
+    }
+    {
+        id: "SDKAppendsGzipAndIgnoresHttpProvidedEncoding_awsJson1_0"
+        documentation: """
+        Compression algorithm encoding is appended to the Content-Encoding header, and the
+        user-provided content-encoding is NOT in the Content-Encoding header since HTTP binding
+        traits are ignored in the awsJson1_0 protocol.
+        """
+        protocol: awsJson1_0
+        params: {
+        "encoding": "custom"
+        "data": """
+                NOTE: REMOVED CONTENTS FOR SAKE OF SPACE. TEST CAN BE FOUND IN REFERENCES SECTION 
+                """
+        }
+        method: "POST"
+        uri: "/"
+        headers: {
+        "Content-Encoding": "gzip"
+        }
+    }
+])
+
+@requestCompression(
+    encodings: ["gzip"]
+)
+operation PutWithContentEncoding {
+        input: PutWithContentEncodingInput
+}
+
+@input
+structure PutWithContentEncodingInput {
+    @httpHeader("Content-Encoding")
+    encoding: String
+
+    data: String
+}
+```
+
+Unit tests for interceptor:
+```kotlin
+@Test
+fun testCorrectHeaderIsAdded() = runTest {...}
+
+@Test
+fun testAlgorithmIsAppendedInHeader() = runTest {...}
+
+@Test
+fun testCompressionIsSkipped() = runTest {...}
+
+@Test
+fun testGzipCompression() = runTest {...}
+
+@Test
+fun testHandlesMultipleCompressionAlgorithms() = runTest {...}
+
+@Test
+fun testNoValidCompressionAlgorithm() = runTest {...}
+```
+
+These tests should pass with no problem.
+
+## Summary
+The request compression trait should be useful for clients that are sending a lot of large requests and could save some bandwidth. What we have right now in this design doc serves as a bare bones implementation of the feature that could eventually be extended to include server to client compression, user supplied compression algorithms, and the like. More compression algorithms could be added as they start being required and it wouldn't be too difficult of a task. It was written with flexibility in mind and should be able to deliver. Some references to things mentioned in passing and assumed to be background knowledge that might be useful are present in the section below.
```

Comment:
**Style**: Closing summaries are generally not necessary in our design specifications. This can be removed.